lib/srv/db/common/permissions/import_rules.go

```diff
@@ -61,39 +65,121 @@ func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObject
 
 	// find all objects that match any of the rules
 	for _, obj := range objs {
-		var objClone *dbobjectv1.DatabaseObject
+		// prepare object clone
+		objClone := utils.CloneProtoMsg(obj)
+		if objClone.Metadata.Labels == nil {
+			objClone.Metadata.Labels = map[string]string{}
+		}
 
 		// apply each mapping in order.
+		matched := false
 		for _, mapping := range mappings {
-			// the matching is applied to the object spec; existing object labels does not matter
-			if !databaseObjectScopeMatch(mapping.GetScope(), obj.GetSpec()) {
-				continue
-			}
-			if databaseObjectImportMatch(mapping.GetMatch(), obj.GetSpec()) {
-				if objClone == nil {
-					objClone = utils.CloneProtoMsg(obj)
-				}
-
-				// mapping applies additional labels
-				labels := objClone.Metadata.Labels
-				if labels == nil {
-					labels = map[string]string{}
-				}
-				for k, v := range mapping.AddLabels {
-					labels[k] = v
-				}
-				objClone.Metadata.Labels = labels
+			if applyMappingToObject(mapping, objClone.GetSpec(), objClone.Metadata.Labels) {
+				matched = true
 			}
 		}
 
-		if objClone != nil {
+		if matched {
 			out = append(out, objClone)
 		}
 	}
 
 	return out
 }
 
+// splitExpressionRE matches the pattern "{{...}}"
+var splitExpressionRE = regexp.MustCompile(`{{[^}]*}}`)
+
+// splitExpressions splits a complex template like `foo-{{bar}}-{{baz}}` into its constituent parts,
+// which can be evaluated one by one.
+func splitExpressions(input string) []string {
+	indexes := splitExpressionRE.FindAllStringIndex(input, -1)
+
+	substringStart := 0
+	parts := make([]string, 0)
+
+	// iterate over matches
+	for _, matchIndex := range indexes {
+		// non-interpolated part
+		parts = append(parts, input[substringStart:matchIndex[0]])
+
+		// interpolated part
+		parts = append(parts, input[matchIndex[0]:matchIndex[1]])
+
+		// move start index
+		substringStart = matchIndex[1]
+	}
+
+	// remaining Append the remaining part of the string after the last match
+	parts = append(parts, input[substringStart:])
+
+	return parts
+}
+
+func evalMultiTemplate(input string, values map[string]string) (string, error) {
+	traits := map[string][]string{}
+	for k, v := range values {
+		traits[k] = []string{v}
+	}
+
+	var sb strings.Builder
+
+	parts := splitExpressions(input)
+	for _, part := range parts {
+		// skip empty parts
+		if part == "" {
+			continue
+		}
+
+		out, err := services.ApplyValueTraits(part, traits)
```

Comment:
there is a TODO in `ApplyValueTraits`
https://github.com/gravitational/teleport/blob/c5529cd5312d5b66fbbf17e492dc59da9934af1c/lib/services/role.go#L632-L640
Should we be worried if that might break here someday?

Reply:
I'll look into the validation bit.

Reply:
I took a deep dive into validation... oh boy. Long story short, this is an actual problem. Unless we are happy with accepting that `{{whatever.object_kind}}` evaluates happily to e.g. `table`, some changes will be needed. I'm trying to force our libraries to behave, but some changes will be needed.

Reply:
All right, this took some work but I've now made some independent code using the underlying library, which doesn't have the problem of mixing up namespaces. Added a few tests on top.